text/0116-sentry-semantic-conventions.md

```diff
@@ -0,0 +1,159 @@
+
+The attributes schema will be versioned and published as a JSON Schema, which will be the representation of the well-known attributes that Sentry ingestion and product relies on.
+
+The JSON schema will also be codegened into packages for JavaScript, Python, and Rust, so that the schema can be used in Relay, the frontend and the backend. There has to be care put in to make sure that the codegened packages are compatible with the JSON schema, and that the JSON schema is backwards compatible, but we can introduce tests to make sure that this is the case.
+
+By having a versionig/packaging structure like this, it also makes it much easier for the SDKs to identify what parts of the product are using the data they are sending.
+
+## Mapping Sentry Specific Fields to Attributes
+
+Right now we have a variety of top level fields that need to be mapped to attributes, and some of them match existing OpenTelemetry semantic conventions. We should map these fields to attributes as follows (the dots in the initial fields represent nested fields):
+
+Top level fields:
+
+- `release` -> `sentry.release`
+- `environment` -> `sentry.environment`
+
+[User interface](https://develop.sentry.dev/sdk/event-payloads/user/):
+
+- `user.id` -> `sentry.user.id`
+- `user.username` -> `sentry.user.username`
+- `user.email` -> `sentry.user.email`
+- `user.ip_address` -> `sentry.user.ip_address`
+- `user.segment` -> `sentry.user.segment`
+- `user.geo.city` -> `sentry.user.geo.city`
+- `user.geo.country_code` -> `sentry.user.geo.country_code`
+- `user.geo.region` -> `sentry.user.geo.region`
+
+[Request interface](https://develop.sentry.dev/sdk/event-payloads/request/)
+
+- `request.method` -> [`http.request.method`](https://github.com/open-telemetry/semantic-conventions/blob/cadfe53949266d33476b15ca52c92f682600a29c/model/trace/http.yaml#L19)
+- `request.url` -> [`url.full`](https://github.com/open-telemetry/semantic-conventions/blob/cadfe53949266d33476b15ca52c92f682600a29c/model/trace/http.yaml#L47)
+- `request.query_string` -> [`url.query_string`](https://github.com/open-telemetry/semantic-conventions/blob/cadfe53949266d33476b15ca52c92f682600a29c/model/trace/http.yaml#L47)
+- `request.cookies` -> `http.request.cookies`
+- `request.cookies` -> `http.request.headers`
+- `request.env` -> `http.request.env`
+
+[SDK interface](https://develop.sentry.dev/sdk/event-payloads/sdk/)
+
+- `sdk.name` -> `sentry.sdk.name`
+- `sdk.version` -> `sentry.sdk.version`
+- `sdk.integrations` -> `sentry.sdk.integrations`
+- `sdk.packages` -> `sentry.sdk.packages`
+
+For the contexts defined in as part of our contexts payload, we can flatten it so that the context name become the prefix to the attribute key. For example the `os` context would become `os.name`, `os.version`, `os.kernel_version`, etc. Some of these new keys will have to be renamed to better match OpenTelemetry's semantic conventions, but that can be approached in a case by case basis.
+
```

Comment:
One other thing that we may have to handle, but maybe also want to do in a follow up, are breadcrumbs? The timed events in OTEL also use the same attributes schema. In the POTEL experiment I have:

* `sentry.breadcrumb.type`
* `sentry.breadcrumb.level`
* `sentry.breadcrumb.event_id`
* `sentry.breadcrumb.category`
* `sentry.breadcrumb.data` --> this is a bit tricky as it is arbitrary JSON, currently I store this as serialized JSON string.

Reply:
Ideally breadcrumbs data matches the attributes, we can structure them the same way otel does it's logs signal.

for span events that are attached that map to breadcrumbs, we probably want to "unroll" breadcrumb data into the top level keys for attributes on span events.

Reply:
to clarify, I do store these as regular attributes on a timed event :) But we still need universally understood attributes on there, I guess? timed events have only a name other than the attributes. Not sure what we do with these fields on the breadcrumb right now though, tbh.